File: python/slhdsa/hypertree.py
"""Hypertree signature for SLH-DSA per FIPS 205, Section 8."""

from slhdsa.params import SLHDSAParams
from slhdsa.address import ADRS
from slhdsa.xmss import xmss_sign, xmss_pk_from_sig, xmss_node


def ht_sign(params: SLHDSAParams, m: bytes, sk_seed: bytes,
            pk_seed: bytes, idx_tree: int, idx_leaf: int) -> bytes:
    """Hypertree signature generation (Algorithm 11 / HT).

    Args:
        m: n-byte message to sign.
        idx_tree: Tree index.
        idx_leaf: Leaf index within the tree.

    Returns:
        Hypertree signature (d XMSS signatures concatenated).
    """
    adrs = ADRS()
    adrs.set_layer_address(0)
    adrs.set_tree_address(idx_tree)

    sig_tmp = xmss_sign(params, m, sk_seed, idx_leaf, pk_seed, adrs)
    sig_ht = sig_tmp

    root = xmss_pk_from_sig(params, idx_leaf, sig_tmp, m, pk_seed, adrs)

    for j in range(1, params.d):
        idx_leaf = idx_tree % (1 << params.hp)
        idx_tree = idx_tree >> params.hp
        adrs.set_layer_address(j)
        adrs.set_tree_address(idx_tree)
        sig_tmp = xmss_sign(params, root, sk_seed, idx_leaf, pk_seed, adrs)
        sig_ht += sig_tmp

        if j < params.d - 1:
            root = xmss_pk_from_sig(params, idx_leaf, sig_tmp, root, pk_seed, adrs)

    return sig_ht


def ht_verify(params: SLHDSAParams, m: bytes, sig_ht: bytes,
              pk_seed: bytes, idx_tree: int, idx_leaf: int,
              pk_root: bytes) -> bool:
    """Hypertree signature verification (Algorithm 12 / HT).

    Returns:
        True if the signature is valid.
    """
    adrs = ADRS()
    xmss_sig_size = (params.wots_len + params.hp) * params.n

    adrs.set_layer_address(0)
    adrs.set_tree_address(idx_tree)

    sig_tmp = sig_ht[:xmss_sig_size]
    node = xmss_pk_from_sig(params, idx_leaf, sig_tmp, m, pk_seed, adrs)

    for j in range(1, params.d):
        idx_leaf = idx_tree % (1 << params.hp)
        idx_tree = idx_tree >> params.hp
        adrs.set_layer_address(j)
        adrs.set_tree_address(idx_tree)
        sig_tmp = sig_ht[j * xmss_sig_size:(j + 1) * xmss_sig_size]
        node = xmss_pk_from_sig(params, idx_leaf, sig_tmp, node, pk_seed, adrs)

    return node == pk_root

File: js/src/slhdsa/hypertree.js
/**
 * Hypertree for SLH-DSA (FIPS 205)
 * Algorithms 16-17
 */

import { ADRS, ADDR_TYPE } from './address.js';
import { xmssNode, xmssSign, xmssPkFromSig } from './xmss.js';
import { concat } from './utils.js';

/**
 * htSign(M, skSeed, pkSeed, idxTree, idxLeaf, params, hashSuite) - Algorithm 16
 * Generate a hypertree signature
 *
 * idxTree: BigInt tree index
 * idxLeaf: integer leaf index
 */
function htSign(M, skSeed, pkSeed, idxTree, idxLeaf, params, hashSuite) {
  const { d, hp, n, len } = params;

  const adrs = new ADRS();
  adrs.setTreeAddress(idxTree);

  // Sign at layer 0
  const sigXmss0 = xmssSign(M, skSeed, idxLeaf, pkSeed, adrs, params, hashSuite);

  const sigParts = [sigXmss0];

  // Compute root at layer 0 for next layer's message
  let root = xmssPkFromSig(idxLeaf, sigXmss0, M, pkSeed, adrs, params, hashSuite);

  // Sign at subsequent layers
  let tree = idxTree;
  for (let j = 1; j < d; j++) {
    const leaf = Number(BigInt(tree) & BigInt((1 << hp) - 1));
    tree = BigInt(tree) >> BigInt(hp);

    adrs.setLayerAddress(j);
    adrs.setTreeAddress(tree);

    const sigJ = xmssSign(root, skSeed, leaf, pkSeed, adrs, params, hashSuite);
    sigParts.push(sigJ);

    if (j < d - 1) {
      root = xmssPkFromSig(leaf, sigJ, root, pkSeed, adrs, params, hashSuite);
    }
  }

  return concat(...sigParts);
}

/**
 * htVerify(M, sigHt, pkSeed, idxTree, idxLeaf, pkRoot, params, hashSuite) - Algorithm 17
 * Verify a hypertree signature
 */
function htVerify(M, sigHt, pkSeed, idxTree, idxLeaf, pkRoot, params, hashSuite) {
  const { d, hp, n, len } = params;
  const xmssSigLen = (len + hp) * n;

  const adrs = new ADRS();
  adrs.setTreeAddress(idxTree);

  // Parse first XMSS signature
  const sig0 = sigHt.slice(0, xmssSigLen);
  let node = xmssPkFromSig(idxLeaf, sig0, M, pkSeed, adrs, params, hashSuite);

  // Verify subsequent layers
  let tree = idxTree;
  for (let j = 1; j < d; j++) {
    const leaf = Number(BigInt(tree) & BigInt((1 << hp) - 1));
    tree = BigInt(tree) >> BigInt(hp);

    adrs.setLayerAddress(j);
    adrs.setTreeAddress(tree);

    const sigJ = sigHt.slice(j * xmssSigLen, (j + 1) * xmssSigLen);
    node = xmssPkFromSig(leaf, sigJ, node, pkSeed, adrs, params, hashSuite);
  }

  // Check if computed root matches pk root
  if (node.length !== pkRoot.length) return false;
  let eq = 0;
  for (let i = 0; i < node.length; i++) {
    eq |= node[i] ^ pkRoot[i];
  }
  return eq === 0;
}

export { htSign, htVerify };
